tests/unittests/test_backoff_error_handling.py

```diff
@@ -0,0 +1,2013 @@
+
+    def GetAdGroupsByCampaignId(self, CampaignId):
+        """ Mocked GetAdGroupsByCampaignId to test the backoff in sync_ad_groups method"""
+        raise self.error
+
+    def GetAdsByAdGroupId(self, AdGroupId, AdTypes):
+        """ Mocked GetAdsByAdGroupId test to the backoff in sync_ads method"""
+        raise self.error
+
+    def SubmitGenerateReport(self, report_request):
+        """ Mocked SubmitGenerateReport to test the backoff in get_report_request_id method"""
+        raise self.error
+
+    def PollGenerateReport(self, request_id):
+        """ Mocked PollGenerateReport to test the backoff in poll_report method"""
+        raise self.error
+
+    @property
+    def factory(self):
+        """ Mocked factory to test backoff the in build_report_request method"""
+        raise self.error
+
+    @property
+    def soap_client(self):
+        """ Mocked soap_client to test backoff in get_type_map method"""
+        raise self.error
+
+@mock.patch("tap_bing_ads.filter_selected_fields_many", return_value = '')
+@mock.patch("singer.write_records", return_value = '')
+@mock.patch("singer.metrics", return_value = '')
+@mock.patch("singer.write_bookmark", return_value = '')
+@mock.patch("singer.write_state", return_value = '')
+@mock.patch("tap_bing_ads.sobject_to_dict", return_value = '')
+@mock.patch("singer.get_bookmark", return_value = {'b1': 'b1'})
+@mock.patch("singer.write_schema", return_value = '')
+@mock.patch("tap_bing_ads.get_core_schema", return_value = '')
+@mock.patch("tap_bing_ads.get_selected_fields", return_value = '')
+class TestBackoffError(unittest.TestCase):
+    '''
+    Test that backoff logic works properly. Mocked some common method to test the backoff including
+    filter_selected_fields_many, write_records , metrics, write_bookmark, write_state, sobject_to_dict,
+    get_bookmark, write_schema, get_core_schema, get_selected_fields, time.sleep.
+    '''
+    @mock.patch("tap_bing_ads.create_sdk_client", return_value = '')
+    @mock.patch("tap_bing_ads.CustomServiceClient")
+    def test_connection_reset_error_get_account(self, mock_get_account, mock_create_sdk_client,
+                                                    mock_get_selected_fields, mock_get_core_schema,
+                                                    mock_write_schema, mock_get_bookmark,
+                                                    mock_sobject_to_dict, mock_write_state,
+                                                    mock_write_bookmark, mock_metrics, mock_write_records,
+                                                    mock_filter_selected_fields_many):
+        '''
+        Test that tap retry for 60 seconds on the connection reset error.
+        '''
+        mock_get_account.side_effect = socket.error(104, 'Connection reset by peer')
+        before_time = datetime.datetime.now()
+        try:
+            tap_bing_ads.sync_accounts_stream(['i1'], {})
+        except ConnectionResetError:
+            pass
```

Comment:
I think this pattern could produce a false negative, if the method were to take more than 60 seconds and succeed rather than raise the specified error. I think this can be fixed by raising an assertion error inside the try after the target method is called.